### PR TITLE
Add ScopedUpdater components to improve re-render performance

### PR DESCRIPTION
**Pull request by cottom**
## Summary

Previously we have meet some cases that interaction in some page is not in time which makes users feel restlessness. The cause is that goji-runtime always update UI in the page scope. While as the page DOM amount increased, time cost  for update every time render can increase exponentially as well. 

This PR is to add a component to limit the update scope manually by developer.

## Usage

we provide `ScopeUpdater` as a react component and it has a corresponding component `scope-updater` in `miniapp-runtime. Every UI update under this component will be sent by `scope-update`  as a host.

```tsx
<Parent>
  <Bala/>
  <ScopeUpdater>
    <Timmer />
  </ScopeUpdater>
  <Balabala/>
<Parent/>
```

## implementation

1. [webpack-plugin] add `scope-updater` component in `goji-runtime`. Add `isHostWrapped` props to describe wrapper component who need to handle events and styles by itself. 
2. [core] add `sendUpdate` to process the `diff`, pickup the host `scopeUpdate` component by update path and update the diff just in the scope.

#### environment
MacBook Pro (16-inch, 2019)
processor 2.4 GHz 8-Core Intel Core i9
Memory 32 GB 2667 MHz DDR4

#### test case

```tsx
export const BenchmarkControl: React.FC<{ isTreatment: boolean; nestCount: number }> = ({ isTreatment, nestCount }) => {
  const [num, setNum] = useState(0);

  const child = isTreatments ? (
    <ScopeUpdater>
      <View>{num}</View>
    </ScopeUpdater>
  ) : (
    <View>{num}</View>
  );

  const loop = (count: number) => {
    return (
      <View className={styles.container}>
        <View>
          <View>{count === 0 ? child : loop(count - 1)}</View>
          <View />
        </View>
      </View>
    );
  };

  return (
    <>
      {loop(nestCount)}
      <Button onTap={() => setNum(i => i + 1)}>++</Button>
    </>
  );
};

```

image

#### prod case (P2 filter)

p2-filter

### CODE REVIEW

**Review by malash — CHANGES_REQUESTED** [review]
Thanks @cottom for your PR. I think it's a good creation to solve update performance in GojiJS. But, there are several problems we have to resolve.

1. Could you please rename the component to `ScopedUpdater` ?
2. Use `sendUpdateData` to transform JSON data is not a good idea. I prefer to modify `pure` function to support this feature.
3. Advice to use `Map<ComponentInstance, Updates>` for `scopedUpdates`
4. Add test cases for this new feature.
5. [useRef](https://goji.js.org/docs/en/refs) may not work inside `ScopedUpdater`.
6. Did you try this component on all platforms ?

File: packages/webpack-plugin/templates/components/scroll-view.js.ejs [hunk 1]
```diff
@@ -8,7 +8,7 @@ Component({
     internalScrollIntoView: '',
   },
   properties: {
-    nodes: {
+    c: {
```
Comment (malash) [inline packages/webpack-plugin/templates/components/scroll-view.js.ejs:11]: Why this change ?

File: packages/core/src/reconciler/instance.ts [hunk 2]
```diff
@@ -48,6 +48,10 @@ export abstract class BaseInstance {
     this.parent = parent;
   }
 
+  public getParent() {
```
Comment (malash) [inline packages/core/src/reconciler/instance.ts:51]: Useless function

File: packages/webpack-plugin/templates/components/scroll-view.js.ejs (hunk 1 above)
Comment (cottom) [reply packages/webpack-plugin/templates/components/scroll-view.js.ejs:11]: unified all `nodes` properties

**Comment by cottom** [comment]
> Thanks @cottom for your PR. I think it's a good creation to solve update performance in GojiJS. But, there are several problems we have to resolve.
> 
> 1. Could you please rename the component to `ScopedUpdater` ?
> 2. Use `sendUpdateData` to transform JSON data is not a good idea. I prefer to modify `pure` function to support this feature.
> 3. Advice to use `Map<ComponentInstance, Updates>` for `scopedUpdates`
> 4. Add test cases for this new feature.
> 5. [useRef](https://goji.js.org/docs/en/refs) may not work inside `ScopedUpdater`.
> 6. Did you try this component on all platforms ?

1. NIT, updated.
2. change pure is not easy to meet the demand of the feature and also break lots of test case. oervall pure is not a good practice here. Just leave it for later refractor.
3. NIT, updated.
4. updated.
5. updated.
6. Yes.

File: packages/webpack-plugin/templates/leaf-components.wxml.ejs [hunk 3]
```diff
@@ -2,16 +2,18 @@
 <% for (var component of components) {%>
   <block wx:elif="{{<% if (component.sid === undefined) {%>type === '<%= component.name %>'<%} else {%>sid === <%= component.sid %><%}%>}}">
     <<%= component.isWrapped ? 'goji-' : ''%><%= component.name %>
-      <% if (component.isWrapped) { %>
+      <% if (component.isWrapped && !component.isSelfHost) { %>
```
Comment (malash) [inline packages/webpack-plugin/templates/leaf-components.wxml.ejs:5]: isHostWrapped

File: packages/core/src/components/scopedUpdater.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+) => {
+  if (ENABLE_SCOPED_UPDATER) {
+    return createElement(TYPE_SCOPE_UPDATER, {
+      ref,
+      ...props
+    });
+  }
+
+  return createElement(Fragment, {
+    ref,
```
Comment (malash) [inline packages/core/src/components/scopedUpdater.tsx:26]: No need to forward ref here

File: packages/core/src/components/scopedUpdater.tsx [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+export type ScopedUpdaterProps = React.PropsWithChildren<{
+  testID?: string;
+  id?: string;
+  className?: string;
+  style?: CSSProperties;
+}>;
+
+const SelfScopedUpdater = (
+  props: ScopedUpdaterProps,
```
Comment (malash) [inline packages/core/src/components/scopedUpdater.tsx:15]: Use `unsafe_className` and `unsafe_style`

File: packages/webpack-plugin/src/constants/components.ts [hunk 6]
```diff
@@ -604,8 +605,21 @@ export const getBuiltInComponents = (target: GojiTarget): ComponentDesc[] =>
     ]),
   );
 
+export const getExtendComponents = (target: GojiTarget): ComponentDesc[] => {
+  return [
```
Comment (malash) [inline packages/webpack-plugin/src/constants/components.ts:609]: `['wechat', 'qq'].includes(target) ? [{ xxx }] : []`

File: packages/webpack-plugin/templates/leaf-components.wxml.ejs [hunk 7]
```diff
@@ -2,16 +2,18 @@
...
         className="{{props.className}}"
         the-id="{{props.id}}"
         the-style="{{props.style || ''}}"
+        c="{{c}}"
       <% } else { %>
         data-goji-id="{{id || -1}}"
         class="{{props.className}}"
         id="{{props.id}}"
         style="{{props.style || ''}}"
+        c="{{c}}"
```
Comment (malash) [inline packages/webpack-plugin/templates/leaf-components.wxml.ejs:16]: Remove this line

**Review by malash — APPROVED** [review]

**Pull request merged**